fabric-transfer-api-v1/src/testmod/java/net/fabricmc/fabric/test/transfer/gametests/VanillaStorageTests.java

```diff
@@ -123,6 +128,85 @@ public void testChestComparator(TestContext context) {
+		try (Transaction transaction = Transaction.openOuter()) {
+			storage.insert(book, 2, transaction);
+
+			if (bookshelf.getStack(0).getCount() != 1) throw new GameTestException("Bookshelf stack 0 should have size 1");
+			if (!book.matches(bookshelf.getStack(0))) throw new GameTestException("Bookshelf stack 0 should be a book");
+			if (bookshelf.getStack(1).getCount() != 1) throw new GameTestException("Bookshelf stack 1 should have size 1");
+			if (!book.matches(bookshelf.getStack(1))) throw new GameTestException("Bookshelf stack 1 should be a book and quill");
+		}
+
+		if (!bookshelf.getStack(0).isEmpty()) throw new GameTestException("Bookshelf stack 0 should be empty again after aborting transaction");
+		if (!bookshelf.getStack(1).isEmpty()) throw new GameTestException("Bookshelf stack 1 should be empty again after aborting transaction");
+
+		// Second, check that we correctly update the last modified slot.
+		try (Transaction tx = Transaction.openOuter()) {
+			if (storage.getSlot(1).insert(book, 1, tx) != 1) throw new GameTestException("Should have inserted 1 book");
+			if (bookshelf.method_47887() != 1) throw new GameTestException("Last modified slot should be 1");
+
+			if (storage.getSlot(2).insert(book, 1, tx) != 1) throw new GameTestException("Should have inserted 1 book");
+			if (bookshelf.method_47887() != 2) throw new GameTestException("Last modified slot should be 2");
+
+			if (storage.getSlot(1).extract(book, 1, tx) != 1) throw new GameTestException("Should have extracted 1 book");
+			if (bookshelf.method_47887() != 1) throw new GameTestException("Last modified slot should be 1");
+
+			// Now, create an aborted nested transaction.
+			try (Transaction nested = tx.openNested()) {
+				if (storage.insert(book, 100, nested) != 5) throw new GameTestException("Should have inserted 5 books");
+				// Now, last modified slot should be 5.
+				if (bookshelf.method_47887() != 5) throw new GameTestException("Last modified slot should be 5");
+			}
+
+			// And it's back to 1 in theory.
+			if (bookshelf.method_47887() != 1) throw new GameTestException("Last modified slot should be 1");
+			tx.commit();
+		}
+
+		if (bookshelf.method_47887() != 1) throw new GameTestException("Last modified slot should be 1 after committing transaction");
+
+		// Let's also check the state properties. Only slot 2 should be occupied.
+		BlockState state = bookshelf.getCachedState();
+
+		if (state.get(Properties.SLOT_0_OCCUPIED)) throw new GameTestException("Slot 0 should not be occupied");
+		if (state.get(Properties.SLOT_1_OCCUPIED)) throw new GameTestException("Slot 1 should not be occupied");
+		if (!state.get(Properties.SLOT_2_OCCUPIED)) throw new GameTestException("Slot 2 should be occupied");
+		if (state.get(Properties.SLOT_3_OCCUPIED)) throw new GameTestException("Slot 3 should be occupied");
+		if (state.get(Properties.SLOT_4_OCCUPIED)) throw new GameTestException("Slot 4 should be occupied");
```

Comment:
Exception text for tests on slots 3 and 4 are inverted

Reply:
Good catch, thank you! :smile: